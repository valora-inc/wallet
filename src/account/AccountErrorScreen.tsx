import React from 'react'
import { StyleSheet, Text } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import TextButton from 'src/components/TextButton'
import fontStyles from 'src/styles/fonts'
import { Spacing } from 'src/styles/styles'

interface Props {
  title: string
  testID: string
  description: string
  onPress(): void
  buttonLabel: string
  onPressSecondary?(): void
  secondaryButtonLabel?: string
}

function AccountErrorScreen({
  title,
  testID,
  description,
  onPress,
  buttonLabel,
  secondaryButtonLabel,
  onPressSecondary,
}: Props) {
  return (
    <SafeAreaView style={styles.content}>
      <Text style={styles.title} testID={testID}>
        {title}
      </Text>
      <Text style={styles.body}>{description}</Text>
      <TextButton onPress={onPress} testID={`${testID}Button`}>
        {buttonLabel}
      </TextButton>
      {secondaryButtonLabel && onPressSecondary && (
        <TextButton
          style={styles.secondaryButton}
          onPress={onPressSecondary}
          testID={`${testID}ButtonSecondary`}
        >
          {secondaryButtonLabel}
        </TextButton>
      )}
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  content: {
    flex: 1,
    marginHorizontal: Spacing.Thick24,
    justifyContent: 'center',
    alignItems: 'center',
  },
  title: {
    ...fontStyles.h2,
    textAlign: 'center',
    paddingBottom: Spacing.Regular16,
    paddingTop: Spacing.Regular16,
  },
  body: {
    ...fontStyles.regular,
    textAlign: 'center',
    paddingBottom: Spacing.Thick24,
  },
  secondaryButton: {
    paddingTop: Spacing.Thick24,
  },
})

export default AccountErrorScreen
