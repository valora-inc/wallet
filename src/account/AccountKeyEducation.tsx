import { StackScreenProps, TransitionPresets } from '@react-navigation/stack'
import React, { useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import Education, { EducationTopic, EmbeddedNavBar } from 'src/account/Education'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { BtnTypes } from 'src/components/Button'
import { accountKey1, accountKey2, accountKey3, accountKey4 } from 'src/images/Images'
import { noHeader } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'

type Props = StackScreenProps<StackParamList, Screens.AccountKeyEducation>

export default function AccountKeyEducation(props: Props) {
  function onComplete() {
    ValoraAnalytics.track(OnboardingEvents.backup_education_complete)
    if (props.route.params?.nextScreen) {
      navigate(props.route.params?.nextScreen)
    } else {
      navigate(Screens.BackupPhrase)
    }
  }

  const { t } = useTranslation()

  const steps = useSteps()

  useEffect(() => {
    ValoraAnalytics.track(OnboardingEvents.backup_education_start)
  }, [])

  return (
    <Education
      embeddedNavBar={EmbeddedNavBar.Close}
      stepInfo={steps}
      onFinish={onComplete}
      finalButtonText={t('completeEducation')}
      buttonText={t('next')}
      finalButtonType={BtnTypes.PRIMARY}
    />
  )
}

AccountKeyEducation.navigationOptions = {
  ...noHeader,
  ...TransitionPresets.ModalTransition,
}

function useSteps() {
  const { t } = useTranslation()
  return React.useMemo(
    () =>
      [
        { image: accountKey1, topic: EducationTopic.backup },
        { image: accountKey2, topic: EducationTopic.backup },
        { image: accountKey3, topic: EducationTopic.backup },
        { image: accountKey4, topic: EducationTopic.backup },
      ].map((step, index) => {
        return {
          ...step,
          title: t(`guide.${index}.title`),
          text: t(`guide.${index}.text`),
        }
      }),
    []
  )
}
