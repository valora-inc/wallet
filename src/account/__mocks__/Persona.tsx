import React from 'react'
import { Button } from 'react-native'

const MockPersona = (props: any) => {
  return (
    <>
      <Button
        title={'mock Persona button'}
        testID="PersonaButton"
        disabled={props.disabled}
        onPress={props.onPress}
      />
      {/* Mock buttons to allow simulating different persona callbacks */}
      <Button
        title={'mock Persona success'}
        testID="PersonaSuccessButton"
        disabled={props.disabled}
        onPress={props.onSuccess}
      />
      <Button
        title={'mock Persona error'}
        testID="PersonaErrorButton"
        disabled={props.disabled}
        onPress={props.onError}
      />
      <Button
        title={'mock Persona cancel'}
        testID="PersonaCancelButton"
        disabled={props.disabled}
        onPress={props.onCanceled}
      />
    </>
  )
}

export default MockPersona
