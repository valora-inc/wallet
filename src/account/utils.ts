import { parsePhoneNumber } from '@celo/phone-utils'

const ADDRESS_LENGTH = 42 // TODO(ACT-1173): see if this can be replaced with a viem helper

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}

export function getPhoneNumberDetails(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      // Show international display number to avoid confusion
      internationalPhoneNumber: phoneDetails.displayNumberInternational,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      internationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
