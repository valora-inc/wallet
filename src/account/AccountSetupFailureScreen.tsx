import React from 'react'
import { useTranslation } from 'react-i18next'
import RNExitApp from 'react-native-exit-app'
import AccountErrorScreen from 'src/account/AccountErrorScreen'
import { noHeaderGestureDisabled } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'

function AccounSetupFailureScreen() {
  const { t } = useTranslation()

  const onPressCloseApp = () => {
    RNExitApp.exitApp()
  }

  const onPressContactSupport = () => {
    navigate(Screens.SupportContact)
  }

  return (
    <AccountErrorScreen
      title={t('accountSetupFailed')}
      testID="AccountSetupFailure"
      description={t('accountSetupFailedDescription')}
      onPress={onPressCloseApp}
      buttonLabel={t('closeApp')}
      onPressSecondary={onPressContactSupport}
      secondaryButtonLabel={t('contactSupport')}
    />
  )
}

AccounSetupFailureScreen.navOptions = noHeaderGestureDisabled

export default AccounSetupFailureScreen
