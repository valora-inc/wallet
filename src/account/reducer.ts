import { Actions, ActionTypes } from 'src/account/actions'
import { Actions as AppActions, ActionTypes as AppActionTypes } from 'src/app/actions'
import { DEV_SETTINGS_ACTIVE_INITIALLY } from 'src/config'
import { deleteKeylessBackupCompleted, keylessBackupCompleted } from 'src/keylessBackup/slice'
import { getRehydratePayload, REHYDRATE, RehydrateAction } from 'src/redux/persist-helper'
import Logger from 'src/utils/Logger'
import { isE164NumberStrict } from 'src/utils/phoneNumbers'
import { Actions as Web3Actions, ActionTypes as Web3ActionTypes } from 'src/web3/actions'

interface State {
  name: string | null
  e164PhoneNumber: string | null
  defaultCountryCode: string | null
  contactDetails: UserContactDetails
  devModeActive: boolean
  devModeClickCount: number
  photosNUXClicked: boolean
  pincodeType: PincodeType
  backupCompleted: boolean // recovery phrase backup
  accountCreationTime: number
  startOnboardingTime?: number
  dismissedGetVerified: boolean
  dismissedGoldEducation: boolean
  acceptedTerms: boolean
  hasMigratedToNewBip39: boolean
  choseToRestoreAccount?: boolean
  profileUploaded?: boolean
  recoveringFromStoreWipe?: boolean
  accountToRecoverFromStoreWipe?: string
  celoEducationCompleted: boolean
  recoveryPhraseInOnboardingStatus: RecoveryPhraseInOnboardingStatus
  cloudBackupCompleted: boolean
}

export enum PincodeType {
  Unset = 'Unset',
  CustomPin = 'CustomPin',
  PhoneAuth = 'PhoneAuth',
}

interface UserContactDetails {
  contactId: string | null
  thumbnailPath: string | null
}

export enum KycStatus {
  Created = 'created',
  Completed = 'completed',
  Failed = 'failed',
  Pending = 'pending',
  Expired = 'expired',
  Approved = 'approved',
  Declined = 'declined',
  NeedsReview = 'needs-review',
  NotCreated = 'not-created',
}

// Maintaining this as it's required for migrations
export enum FinclusiveKycStatus {
  NotSubmitted = 0,
  Submitted = 1,
  Accepted = 2,
  Rejected = 3,
  InReview = 4,
}

export enum RecoveryPhraseInOnboardingStatus {
  NotStarted = 'NotStarted', // users who have not reached the "ProtectWallet" screen
  InProgress = 'InProgress', // users who have reached the "ProtectWallet" screen, but not clicked "I've saved it"
  Completed = 'Completed', // users who have clicked "I've saved it"
}

const initialState: State = {
  name: null,
  e164PhoneNumber: null,
  defaultCountryCode: null,
  contactDetails: {
    contactId: null,
    thumbnailPath: null,
  },
  devModeActive: DEV_SETTINGS_ACTIVE_INITIALLY,
  devModeClickCount: 0,
  photosNUXClicked: false,
  pincodeType: PincodeType.Unset,
  accountCreationTime: 99999999999999,
  startOnboardingTime: undefined,
  backupCompleted: false,
  dismissedGetVerified: false,
  dismissedGoldEducation: false,
  acceptedTerms: false,
  hasMigratedToNewBip39: false,
  choseToRestoreAccount: false,
  profileUploaded: false,
  recoveringFromStoreWipe: false,
  accountToRecoverFromStoreWipe: undefined,
  celoEducationCompleted: false,
  recoveryPhraseInOnboardingStatus: RecoveryPhraseInOnboardingStatus.NotStarted,
  cloudBackupCompleted: false,
}

export const reducer = (
  state: State | undefined = initialState,
  action:
    | ActionTypes
    | RehydrateAction
    | Web3ActionTypes
    | AppActionTypes
    | typeof keylessBackupCompleted
    | typeof deleteKeylessBackupCompleted
): State => {
  switch (action.type) {
    case REHYDRATE: {
      const rehydratedPayload = getRehydratePayload(action, 'account')
      // Ignore some persisted properties
      return {
        ...state,
        ...rehydratedPayload,
        dismissedGetVerified: false,
      }
    }
    case Actions.CHOOSE_CREATE_ACCOUNT:
      return {
        ...state,
        choseToRestoreAccount: false,
        startOnboardingTime: state?.startOnboardingTime ?? action.now,
      }
    case Actions.CHOOSE_RESTORE_ACCOUNT:
      return {
        ...state,
        choseToRestoreAccount: true,
      }
    case Actions.START_STORE_WIPE_RECOVERY:
      return {
        ...state,
        choseToRestoreAccount: true,
        recoveringFromStoreWipe: true,
        accountToRecoverFromStoreWipe: action.accountToRecover,
        pincodeType: PincodeType.CustomPin,
        acceptedTerms: true,
      }
    case Actions.CANCEL_CREATE_OR_RESTORE_ACCOUNT:
      return {
        ...state,
        choseToRestoreAccount: false,
        recoveringFromStoreWipe: false,
        pincodeType: PincodeType.Unset,
      }
    case Actions.SET_NAME:
      return {
        ...state,
        name: action.name,
      }
    case Actions.SAVE_NAME:
      return {
        ...state,
        name: action.name,
      }
    case AppActions.PHONE_NUMBER_VERIFICATION_COMPLETED:
    case Actions.SET_PHONE_NUMBER:
      if (!isE164NumberStrict(action.e164PhoneNumber)) {
        return state
      }
      return {
        ...state,
        e164PhoneNumber: action.e164PhoneNumber,
        defaultCountryCode: action.countryCode,
      }
    case AppActions.PHONE_NUMBER_REVOKED:
      return {
        ...state,
        e164PhoneNumber: null,
        defaultCountryCode: null,
      }
    case Actions.DEV_MODE_TRIGGER_CLICKED:
      const newClickCount = (state.devModeClickCount + 1) % 10
      if (newClickCount === 5) {
        Logger.showMessage('Debug Mode Activated')
      } else if (newClickCount === 0) {
        Logger.showMessage('Debug Mode Deactivated')
      }
      return {
        ...state,
        devModeClickCount: newClickCount,
        devModeActive: newClickCount >= 5,
      }
    case Actions.PHOTOSNUX_CLICKED:
      return {
        ...state,
        photosNUXClicked: true,
      }
    case Actions.SET_PINCODE_SUCCESS:
      return {
        ...state,
        pincodeType: action.pincodeType,
      }
    case Actions.SET_PINCODE_FAILURE:
      return {
        ...state,
        pincodeType: PincodeType.Unset,
      }
    case Actions.SET_ACCOUNT_CREATION_TIME:
      return {
        ...state,
        accountCreationTime: action.now,
      }
    case Actions.SET_BACKUP_COMPLETED:
      return {
        ...state,
        backupCompleted: true,
      }
    case Actions.DISMISS_GET_VERIFIED:
      return {
        ...state,
        dismissedGetVerified: true,
      }
    case Actions.DISMISS_GOLD_EDUCATION:
      return {
        ...state,
        dismissedGoldEducation: true,
      }
    case Actions.SET_USER_CONTACT_DETAILS:
      return {
        ...state,
        contactDetails: {
          contactId: action.contactId,
          thumbnailPath: action.thumbnailPath,
        },
      }
    case Actions.ACCEPT_TERMS: {
      return { ...state, acceptedTerms: true }
    }
    case Web3Actions.SET_ACCOUNT: {
      return {
        ...state,
        hasMigratedToNewBip39: true,
      }
    }
    case Actions.PROFILE_UPLOADED: {
      return {
        ...state,
        profileUploaded: true,
      }
    }
    case Actions.SET_CELO_EDUCATION_COMPLETED:
      return {
        ...state,
        celoEducationCompleted: action.celoEducationCompleted,
      }
    case Actions.RECOVERY_PHRASE_IN_ONBOARDING_STARTED:
      return {
        ...state,
        recoveryPhraseInOnboardingStatus: RecoveryPhraseInOnboardingStatus.InProgress,
      }
    case Actions.RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED:
      return {
        ...state,
        recoveryPhraseInOnboardingStatus: RecoveryPhraseInOnboardingStatus.Completed,
      }
    case keylessBackupCompleted.type:
      return {
        ...state,
        cloudBackupCompleted: true,
      }
    case deleteKeylessBackupCompleted.type:
      return {
        ...state,
        cloudBackupCompleted: false,
      }
    default:
      return state
  }
}
