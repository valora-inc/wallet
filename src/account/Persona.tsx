import * as React from 'react'
import { useCallback, useState } from 'react'
import { useAsync } from 'react-async-hook'
import { useTranslation } from 'react-i18next'
import Inquiry, { InquiryAttributes } from 'react-native-persona'
import { useDispatch, useSelector } from 'react-redux'
import { KycStatus } from 'src/account/reducer'
import { showError } from 'src/alert/actions'
import { CICOEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import Button, { BtnSizes, BtnTypes } from 'src/components/Button'
import { readOnceFromFirebase } from 'src/firebase/firebase'
import networkConfig from 'src/geth/networkConfig'
import { createPersonaAccount, verifyWalletAddress } from 'src/in-house-liquidity'
import Logger from 'src/utils/Logger'
import { walletAddressSelector } from 'src/web3/selectors'
// eslint-disable-next-line import/no-relative-packages
import pjson from '../../package.json'

const TAG = 'PERSONA'

export interface Props {
  kycStatus: KycStatus | undefined
  text?: string | undefined
  onPress?: () => any
  onCancelled?: () => any
  onError?: () => any
  onSuccess?: (address: InquiryAttributes['address']) => any
}

const Persona = ({ kycStatus, text, onCancelled, onError, onPress, onSuccess }: Props) => {
  const { t } = useTranslation()
  const [personaAccountCreated, setPersonaAccountCreated] = useState(!!kycStatus)

  const walletAddress = useSelector(walletAddressSelector)

  const dispatch = useDispatch()

  const templateIdResponse = useAsync(async () => readOnceFromFirebase('persona/templateId'), [])
  const templateId = templateIdResponse.result

  const launchPersonaInquiry = useCallback(() => {
    if (typeof templateId !== 'string') {
      Logger.error(TAG, `Attempted to initiate Persona with invalid templateId: ${templateId}`)
      return
    }

    if (!walletAddress) {
      // accountMTWAddress can be null if user's phone number is not verified.
      // Current plan for initial PFP release is to monitor drop off rate for users who haven't verified their phone numbers yet
      // Discussion -> https://valora-app.slack.com/archives/C025V1D6F3J/p1637606953112000
      Logger.warn(TAG, "Can't render Persona because accountMTWAddress is null")
      return
    }
    onPress?.()
    Inquiry.fromTemplate(templateId)
      .referenceId(walletAddress)
      .environment(networkConfig.personaEnvironment)
      .iosTheme(pjson.persona.iosTheme)
      .onSuccess((inquiryId: string, attributes: InquiryAttributes) => {
        onSuccess?.(attributes?.address)
        ValoraAnalytics.track(CICOEvents.persona_kyc_success)
        Logger.info(TAG, `Inquiry completed for ${inquiryId}`)
      })
      .onCancelled(() => {
        onCancelled?.()
        ValoraAnalytics.track(CICOEvents.persona_kyc_cancel)
        Logger.info(TAG, 'Inquiry is canceled by the user.')
      })
      .onError((error: Error) => {
        onError?.()
        ValoraAnalytics.track(CICOEvents.persona_kyc_error)
        Logger.error(TAG, `Error: ${error.message}`)
      })
      .build()
      .start()
  }, [templateId])

  useAsync(async () => {
    if (!personaAccountCreated) {
      try {
        await createPersonaAccount(verifyWalletAddress({ walletAddress }))
        setPersonaAccountCreated(true)
      } catch (error) {
        Logger.warn(TAG, error)
        dispatch(showError(ErrorMessages.PERSONA_ACCOUNT_ENDPOINT_FAIL))
      }
    }
  }, [])

  return (
    <Button
      onPress={launchPersonaInquiry}
      text={text || t('raiseLimitBegin')}
      type={BtnTypes.PRIMARY}
      size={BtnSizes.MEDIUM}
      testID="PersonaButton"
      disabled={!personaAccountCreated || !templateId}
    />
  )
}

export default Persona
