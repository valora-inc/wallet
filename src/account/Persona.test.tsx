import { fireEvent, render, waitFor } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import Inquiry from 'react-native-persona'
import { Provider } from 'react-redux'
import Persona, { Props } from 'src/account/Persona'
import { KycStatus } from 'src/account/reducer'
import { createMockStore } from 'test/utils'
import { mockAccount, mockPrivateDEK } from 'test/values'
import { createPersonaAccount } from 'src/in-house-liquidity'

const FAKE_TEMPLATE_ID = 'fake template id'
jest.mock('react-native-persona')
jest.mock('src/firebase/firebase', () => ({
  readOnceFromFirebase: jest.fn(() => FAKE_TEMPLATE_ID),
}))

jest.mock('src/in-house-liquidity', () => ({
  ...(jest.requireActual('src/in-house-liquidity') as any),
  createPersonaAccount: jest.fn(() => Promise.resolve()),
}))

const mockInquiryBuilder = {
  fromTemplate: jest.fn().mockReturnThis(),
  referenceId: jest.fn().mockReturnThis(),
  environment: jest.fn().mockReturnThis(),
  iosTheme: jest.fn().mockReturnThis(),
  onSuccess: jest.fn().mockReturnThis(),
  onCancelled: jest.fn().mockReturnThis(),
  onError: jest.fn().mockReturnThis(),
  build: jest.fn().mockReturnThis(),
  start: jest.fn().mockReturnThis(),
}
//@ts-ignore Persona doesn't expose the types to cast this mock adequately :\
jest.spyOn(Inquiry, 'fromTemplate').mockReturnValue(mockInquiryBuilder)

describe('Persona', () => {
  const store = createMockStore({
    web3: {
      mtwAddress: mockAccount,
      dataEncryptionKey: mockPrivateDEK,
    },
  })

  beforeEach(() => {
    jest.useRealTimers()
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    const personaProps: Props = {
      kycStatus: KycStatus.Created,
    }

    const { toJSON } = render(
      <Provider store={store}>
        <Persona {...personaProps} />
      </Provider>
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('calls IHL to create a persona account if launching the first time', async () => {
    const personaProps: Props = {
      kycStatus: undefined,
    }
    const { getByTestId } = render(
      <Provider store={store}>
        <Persona {...personaProps} />
      </Provider>
    )
    // Should be disabled to start because we don't know if they have an account until the IHL call happens
    expect(getByTestId('PersonaButton')).toBeDisabled()

    await waitFor(() => expect(getByTestId('PersonaButton')).not.toBeDisabled())
    expect(createPersonaAccount).toHaveBeenCalledTimes(1)
  })

  it('launches persona on button press', async () => {
    const personaProps: Props = {
      kycStatus: KycStatus.Created,
    }
    const { getByTestId } = render(
      <Provider store={store}>
        <Persona {...personaProps} />
      </Provider>
    )

    await waitFor(() => expect(getByTestId('PersonaButton')).not.toBeDisabled())

    fireEvent.press(getByTestId('PersonaButton'))
    expect(Inquiry.fromTemplate).toHaveBeenCalledWith(FAKE_TEMPLATE_ID)
  })

  it('calls onSuccess callback on inquiry success', async () => {
    const personaProps: Props = {
      kycStatus: KycStatus.Created,
      onSuccess: jest.fn(),
    }
    const { getByTestId } = render(
      <Provider store={store}>
        <Persona {...personaProps} />
      </Provider>
    )

    await waitFor(() => expect(getByTestId('PersonaButton')).not.toBeDisabled())

    fireEvent.press(getByTestId('PersonaButton'))
    expect(Inquiry.fromTemplate).toHaveBeenCalledWith(FAKE_TEMPLATE_ID)

    expect(mockInquiryBuilder.onSuccess).toHaveBeenCalled()
    expect(personaProps.onSuccess).not.toHaveBeenCalled()
    mockInquiryBuilder.onSuccess.mock.calls?.[0]?.[0]?.() // simulate Persona invoking the onSuccess callback
    expect(personaProps.onSuccess).toHaveBeenCalled()
  })
  it('calls onCancelled callback on inquiry cancel', async () => {
    const personaProps: Props = {
      kycStatus: KycStatus.Created,
      onCancelled: jest.fn(),
    }
    const { getByTestId } = render(
      <Provider store={store}>
        <Persona {...personaProps} />
      </Provider>
    )

    await waitFor(() => expect(getByTestId('PersonaButton')).not.toBeDisabled())

    fireEvent.press(getByTestId('PersonaButton'))
    expect(Inquiry.fromTemplate).toHaveBeenCalledWith(FAKE_TEMPLATE_ID)

    expect(mockInquiryBuilder.onCancelled).toHaveBeenCalled()
    expect(personaProps.onCancelled).not.toHaveBeenCalled()
    mockInquiryBuilder.onCancelled.mock.calls?.[0]?.[0]?.() // simulate Persona invoking the onCancelled callback
    expect(personaProps.onCancelled).toHaveBeenCalled()
  })
  it('calls onError callback on inquiry error', async () => {
    const personaProps: Props = {
      kycStatus: KycStatus.Created,
      onError: jest.fn(),
    }
    const { getByTestId } = render(
      <Provider store={store}>
        <Persona {...personaProps} />
      </Provider>
    )

    await waitFor(() => expect(getByTestId('PersonaButton')).not.toBeDisabled())

    fireEvent.press(getByTestId('PersonaButton'))
    expect(Inquiry.fromTemplate).toHaveBeenCalledWith(FAKE_TEMPLATE_ID)

    expect(mockInquiryBuilder.onError).toHaveBeenCalled()
    expect(personaProps.onError).not.toHaveBeenCalled()
    mockInquiryBuilder.onError.mock.calls?.[0]?.[0]?.({ message: 'error' }) // simulate Persona invoking the onError callback
    expect(personaProps.onError).toHaveBeenCalled()
  })
})
