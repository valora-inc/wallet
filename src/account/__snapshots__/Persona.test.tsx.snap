// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Persona renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 100,
        },
      ]
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={false}
      nativeBackgroundAndroid={
        Object {
          "attribute": "selectableItemBackground",
          "rippleRadius": undefined,
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#97DFC1",
          "height": 48,
          "justifyContent": "center",
          "minWidth": 120,
          "opacity": undefined,
          "paddingHorizontal": 24,
          "paddingVertical": 5,
        }
      }
      testID="PersonaButton"
    >
      <Text
        maxFontSizeMultiplier={1}
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 16,
            "lineHeight": 22,
          }
        }
      >
        raiseLimitBegin
      </Text>
    </View>
  </View>
</View>
`;
